File: python/2020/Day1.py
import sys
from common import DataAnalyzer


def calc3(data, result=2020):
    length = len(data)
    for i in range(0, length-1):
        s = set()
        total = result - data[i]
        for j in range(i + 1, length):
            if (total - data[j]) in s:
                return i, j, data.index(total - data[j])
            s.add(data[j])
    
    return -1, -1, -1


def calc2(data, result=2020):
    length = len(data)
    for i in range(length):
        for j in range(i, length):
            if i != j and data[i] + data[j] == result:
                return i, j
    
    return -1, -1


def second():
    data = DataAnalyzer.ints("2020/day1.txt")
    x, y, z = calc3(data)
    print("(1.2) 2020 {:d} {:d} {:d} => {:d}".format(x, y, z, data[x]*data[y]*data[z]))


def first():
    values = DataAnalyzer.ints("2020/day1.txt")
    x, y = calc2(values)
    print("(1.1) 2020 {:d} {:d} => {:d}".format(x, y, values[x]*values[y]))


def solve(puzzle):
    if puzzle == '1':
        first()
    elif puzzle == '2':
        second()
    else:
        first()
        second()


if __name__ == '__main__':
    solve(sys.argv[1])

File: js/2020/Day1.js
const {DataAnalyzer} = require('../common/DataAnalyzer.js')
let data = DataAnalyzer.ints("2020/day1.txt")

function calc2(result=2020) {
    let set = new Set()
    set.add(data.shift())

    for (let i = 0; i <= data.length; i++) {
        if (set.has(result - data[i])) {
            return [data[i], result - data[i]]
        }

        set.add(data[i])
    }

    return [-1, -1]
}

function calc3(result=2020) {
    for (let i = 0; i <= data.length - 1; i++) {
        let set = new Set()
        total = result - data[i]

        for (let j = i + 1; j < data.length; j++) {
            if (set.has(total - data[j])) {
                return [data[i], data[j], total - data[j]]
            }

            set.add(data[j])
        }
    }

    return [-1, -1, -1]
}

function one() {
    answer = calc2()
    console.log("(2020.1): " + answer[0] * answer[1])
}

function two() {
    answer = calc3()
    console.log("(2020.2): " + answer[0] * answer[1] * answer[2])
}


switch(process.argv.slice(2)[0]) {
    case '1':
        one();
        break;

    case '2':
        two();
        break;
    
    default:
        one();
        two();
        break;
}